.pre-commit-config.yaml

```diff
@@ -96,6 +96,15 @@ repos:
       - id: yamllint
         files: ^tmt/schemas/.*\.yaml
 
+  - repo: https://github.com/ansible-community/ansible-lint.git
+    rev: v5.3.2
```

Comment:
v6.22.2 is latest. It probably won't need ansible-core version explicitly set here either.

Reply:
I agree, should be the latest version

Reply:
`rev: v5.3.2` is removed.

Reply:
Hi @thrix and @martinhoyer, if we use the latest revision (i.e. `rev: v6.22.2`), there are `303 failures` should be fixed :-) [Here][01] is the output for details, let me fix the yaml files one by one ...

[01]: https://github.com/idorax/tmt-dev-sandbox/blob/dev.20240202.ansible.lint/stub/out05.md

Reply:
You probably want to make sure only ansible yaml files are being checked, not all yaml/yml.   
You can use `files` and/or `exclude` for this: https://pre-commit.com/#regular-expressions

Reply:
Some of the failures with ansible yamls formatting will probably be same or similar as described here: https://github.com/teemtee/tmt/pull/2187

Reply:
@happz what would you say would be a good way to include/exclude ansible files? Something like search for yaml/yml in `tmt/`, `examples/` and perhaps `tests/` but exclude `/test_data/`?  
I'm not super sure what yamls are actually being used with ansible to be honest :)

Reply:
I'd start with directories we know are holding Ansible playbooks. Maybe even just the one with `prepare/feature` playbooks.

I guess we could collect all of them and ignore non-playbook ones, but since this PR slowly turns into "introduce ansible-lint and fix `prepare/feature` ones", my two cents speak against including more into this PR than there already is. I think there's more work needed to consider Ansible playbooks linted and nice, e.g. renaming some of them to fix extension, so an extra PR would be nice, with less `prepare/feature` focus. And I can imagine we'd there switch the approach, i.e. fix them all & enable `ansible-lint` for everything except know non-playbooks :)

Reply:
Well that's what I'm asking for - what are the directories known to hold playbooks, so we can properly filter out only the files we want to check so this PR doesn't need to grow.

With the changes already in this PR, the whole `tmt/` dir passes the "production" lint profile. Well, with a 2 very minor fixes.  
Maybe we could just do that for now?

Going forwards, `tests/` dir passes "min" profile when test_data is excluded and have only 6 "shared" profile fails. 
`examples/` is a bit more tricky, but also not too terrible.

Reply:
```
./ansible/packages.yml
./examples/ansible/test.yml
./examples/redis/ansible/setup_server.yml
./examples/redis/ansible/tasks/redis.yml
./examples/redis/ansible/tasks/redis_variables.yml
./tests/finish/ansible/data/playbook.yml
./tests/prepare/ansible/data/playbook.yml
./tests/run/worktree/data/ansible/playbook.yml

./tmt/steps/prepare/feature/epel-disable.yaml
./tmt/steps/prepare/feature/epel-enable.yaml
```

That seems to be all, every other YAML file is not a playbook. I'd say we shouldn't give `ansible-lint` a free hand across whole top-level directories like `tmt/`, out of 34 YAML files, only 2 are playbooks. The rest might easily violate rules Ansible establishes for their control files.

Reply:
Thanks!  
@idorax I would propose to do:
```
  - repo: https://github.com/ansible-community/ansible-lint.git
    rev: v6.22.2
    hooks:
      - id: ansible-lint
        files: |
          ^(?:.*\/)?ansible\/|
          ^tmt/steps/prepare/feature/
        args: ['--profile=production', '--skip-list=yaml[indentation]', '--exclude=examples/redis/ansible/tasks/redis.yml']
        # redis.yml example requires posix module to be installed in order to be parsed by ansible-lint
```
And go from there.

Reply:
oh, I see: https://github.com/ansible/ansible-lint/blob/36ef08b965cb63a8a7d1efafc45cd569e1a4f87f/.pre-commit-hooks.yaml#L11 
Back to drawing board. Looks like we'll have to use ansible-lint arguments to filter files.

Reply:
So I can match the glob patterns with ansible-lint cli `ansible-lint ansible */ansible */*/ansible/ tmt/steps/prepare/feature --exclude=examples/redis/ansible/tasks/redis.yml`, but am unable to make it work with pre-commit, except by using bash entry:
```
  - repo: https://github.com/ansible-community/ansible-lint.git
    rev: v6.22.2
    hooks:
      - id: ansible-lint
        entry: bash
        args: [
          '-c', 'ansible-lint ansible */ansible */*/ansible tmt/steps/prepare/feature --profile shared 
          --exclude=examples/redis/ansible/tasks/redis.yml']
          # redis.yml example requires posix module to be installed in order to be parsed by ansible-lint
```

Reply:
Or just explicitly set all the ansible dirs. Don't know what is preferred. 
```
  - repo: https://github.com/ansible-community/ansible-lint.git
    rev: v6.22.2
    hooks:
      - id: ansible-lint
        args:
          - ansible/
          - examples/ansible/
          - examples/redis/ansible/
          - tests/finish/ansible/data/
          - tests/run/worktree/data/ansible/
          - tmt/steps/prepare/feature/
          - --profile=production
          - --warn-list=fqcn[action-core]
          - --exclude=examples/redis/ansible/tasks/redis.yml
          # redis.yml example requires posix module to be installed in order to be parsed by ansible-lint
```
The `fqcn[action-core]` causes 10 failures from remaining 24. We can either make it a warning, and not fail like this. Or just fix it and remove the `--warn-list` argument.

Reply:
maybe I would go with an explicit list personally ... seems more clear ... but yeah, ideal would be if it would be able to detect ansible files ;(

Reply:
>  ideal would be if it would be able to detect ansible files ;(

I just had a try to detect ansible files via `ansible-playbook --syntax-check <YAML file>` then check the return code. But it seems not reliable. e.g.

```bash
# ================ foo.sh =====================
#!/bin/bash
NAME=$(basename $(readlink -f $BASH_SOURCE))
TMPDIR=${TMPDIR:-"/tmp"}

verbose=${VERBOSE:-"false"}
output_file=$TMPDIR/$NAME.out.$$
src_root=${TMT_SRC_ROOT:-"$(echo ~/dev/2664/dev)"}
cd $src_root
for file in $(find . -name "*.yaml" -or -name "*.yml"); do
	echo ">>> $file" >> $output_file
	ansible-playbook --syntax-check $file >> $output_file 2>&1
	if (( $? == 0 )); then
		echo "Ansible playbook: ${file#./}"
	else
		[[ $verbose == "true" ]] && echo "Regular YMAL file: ${file#./}"
	fi
done

(tmt) huanli@KVM:2664$ export TMT_SRC_ROOT=/home/huanli/dev/2664/readonly
(tmt) huanli@KVM:2664$ ./foo.sh 
Ansible playbook: ansible/packages.yml
Ansible playbook: examples/ansible/test.yml
Ansible playbook: examples/redis/ansible/setup_server.yml
Ansible playbook: tests/finish/ansible/data/playbook.yml
Ansible playbook: tests/prepare/ansible/data/playbook.yml
Ansible playbook: tests/run/worktree/data/ansible/playbook.yml
Ansible playbook: tmt/steps/prepare/feature/epel-disable.yaml
Ansible playbook: tmt/steps/prepare/feature/epel-enable.yaml
```

File [examples/redis/ansible/tasks/redis.yml][01] and [examples/redis/ansible/tasks/redis_variables.yml][02] are ansible playbooks as a matter of fact, but `ansible-playbook --syntax-check ...` doesn't work as expected. e.g.

```bash
(tmt) huanli@KVM:readonly$ ansible-playbook --syntax-check examples/redis/ansible/tasks/redis.yml 
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'
ERROR! 'package' is not a valid attribute for a Play

The error appears to be in '/home/huanli/dev/2664/readonly/examples/redis/ansible/tasks/redis.yml': line 1, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: setup syncing redis server
  ^ here
(tmt) huanli@KVM:readonly$ echo $?
4
```

[01]: https://github.com/teemtee/tmt/blob/main/examples/redis/ansible/tasks/redis.yml
[02]: https://github.com/teemtee/tmt/blob/main/examples/redis/ansible/tasks/redis_variables.yml

Reply:
@idorax those files are not playbooks, they are files with Ansible tasks. Checking `examples/redis/ansible/setup_server.yml` is sufficient

Reply:
we decided to go with a static list.